cmd/bucket-lifecycle.go

```diff
@@ -218,26 +213,31 @@ func deleteTransitionedObject(ctx context.Context, objectAPI ObjectLayer, bucket
 		// from the source, while leaving metadata behind. The data on
 		// transitioned tier lies untouched and still accessible
 		opts.Transition.Status = lcOpts.TransitionStatus
-		_, err = objectAPI.DeleteObject(ctx, bucket, object, opts)
+		_, err := objectAPI.DeleteObject(ctx, bucket, object, opts)
 		return err
 	}
 
-	// When an object is past expiry, delete the data from transitioned tier and
-	// metadata from source
-	// When an objectglobalTierConfigMgr transitioned tier and
-	// metadata from source
-	if err := tgtClient.Remove(GlobalContext, tgtObjName); err != nil {
-		logger.LogIf(ctx, err)
+	// Request to delete transitioned object comes from DeleteObject API or
+	// an expiry lifecycle event. In case of a DeleteObject API, simply
+	// delete the transitioned object in the remote tier. In case of an
+	// expiry event we should also remove the object metadata in source.
+	tgtClient, err := globalTierConfigMgr.GetDriver(transitionSC)
+	if err != nil {
 		return err
 	}
 
-	// Delete metadata on source, now that transition tier has been cleaned up.
-	if _, err = objectAPI.DeleteObject(ctx, bucket, object, opts); err != nil {
+	if err = tgtClient.Remove(GlobalContext, tgtObjName); err != nil {
 		logger.LogIf(ctx, err)
 		return err
 	}
 
 	if expiryEvent {
+		// Delete metadata on source, now that transition tier has been cleaned up.
```

Comment:
expiryEvent will be true only if it is a lifecycle event - we need to clean up metadata if expiry or `mc rm` called on the transitioned object. That is why it is outside the `if expiryEvent` check